crates/rome_js_analyze/src/analyzers/nursery/no_fallthrough_switch_clause.rs

```diff
@@ -47,36 +59,142 @@ declare_rule! {
     type Options = ();
 
     fn run(ctx: &RuleContext<Self>) -> Self::Signals {
-        let query = ctx.query();
-        let mut signals: Self::Signals = Vec::new();
-        let mut cases = query.cases().into_iter().peekable();
-
-        while let Some(any_case) = cases.next() {
-            let is_last = cases.peek().is_none();
-
-            if is_last {
+        let cfg = ctx.query();
+        let mut fallthrough: Vec<TextRange> = vec![];
+        // Return early if the graph doesn't contain any switch statements.
+        // This avoids to allocate some memory.
+        if !has_switch_statement(&cfg.node) {
+            return fallthrough;
+        }
+        // block to process.
+        let mut block_stack = Vec::new();
+        let mut visited_blocks = FxHashSet::default();
+        block_stack.push(0u32);
+        visited_blocks.insert(0u32);
+        // Traverse the control flow graph and search for switch statements.
+        while let Some(block_index) = block_stack.pop() {
+            // SAFETY: this is a safe conversion because it is already an index for `cfg.blocks`.
+            let block_index = block_index as usize;
+            let Some(block) = cfg.blocks.get(block_index) else {
                 continue;
+            };
+            // Register exception handlers as blocks to process
+            // Ignore finally handler: they are already in the Control Flow Graph.
+            for exception_handler in block
+                .exception_handlers
+                .iter()
+                .filter(|x| matches!(x.kind, ExceptionHandlerKind::Catch))
+            {
+                // Avoid cycles and redundant checks.
+                if visited_blocks.insert(exception_handler.target) {
+                    block_stack.push(exception_handler.target);
+                }
             }
-
-            if case_fell(&any_case) {
-                signals.push(any_case);
+            // Traverse the instructions of the block searching for a switch statement.
+            // A switch statements is followed by conditional jumps (the cases),
+            // and a last unconditional jump (maybe the default clause)
+            let mut is_switch = false;
+            let mut switch_clauses = VecDeque::new();
+            let mut has_default_clause = false;
+            let mut switch_clause_blocks = FxHashSet::default();
+            let mut block_to_switch_clause_range = FxHashMap::default();
+            for instruction in block.instructions.iter() {
+                match instruction.kind {
+                    InstructionKind::Statement => {
+                        if let Some(node) = &instruction.node {
+                            if let Some(switch_stmt) =
+                                node.parent().and_then(JsSwitchStatement::cast)
+                            {
+                                if is_switch {
+                                    unreachable!("A block cannot contain two switch statements.")
```

Comment:
Is there some user code that could hit this code path?

Reply:
It is an assumption about the CFG implementation (every switch get its own block). If we hit this code path it is likely because the CFG implementation was modified.

Reply:
Perfect, is it possible to change the description and try to give a hint to the developer about how to fix the issue? If someday a core contributor stumbles here, we should leave as much info as possible. If we're not able to do so, maybe it's a hint of code that could've been written differently/more efficiently.